Classes/Updater.lua

```diff
@@ -11,129 +11,74 @@ cUpdater = {}
-	end,
-	
-	OnError = function(a_TCPLink, a_ErrorCode, a_ErrorMsg)
-		if (g_DownloadAttempts < g_Config.Updates.NumAttempts) then
-			g_DownloadAttempts = g_DownloadAttempts + 1
-			cUpdater:DownloadLatestVersion()
-		else
-			LOGWARNING("Error while downloading newer worldedit version: " .. a_ErrorCode)
-		end
-	end,
-	
-	OnReceivedData = function(a_TCPLink, a_Data)
-		-- Append the received Data to the saved data
-		g_WorldEditZip = g_WorldEditZip .. a_Data
-	end,
+function cUpdater:CheckForNewerVersion()
+	if (cUpdater.s_CheckAttempts > g_Config.Updates.NumAttempts) then
+		LOGWARNING("Could not connect to Github to check for a newer version for WorldEdit")
+		return;
+	end
+	cUpdater.s_CheckAttempts = cUpdater.s_CheckAttempts + 1;
 	
-	OnRemoteClosed = function(a_TCPLink)
-		-- Split the data and the headers
-		local Header, Data = g_WorldEditZip:match("^(.-)\r\n\r\n(.*)")
-		
-		-- Parse the headers so we can actually use them to validate the data
-		Header = cUpdater:ParseHeaders(Header)
-		
-		-- Check if the received data size matches that that was expected
-		if (Data:len() ~= tonumber(Header["content-length"] or -1)) then
-			if (g_DownloadAttempts < g_Config.Updates.NumAttempts) then
-				g_DownloadAttempts = g_DownloadAttempts + 1
-				cUpdater:DownloadLatestVersion()
+	cUrlClient:Get("https://raw.githubusercontent.com/cuberite/WorldEdit/master/Info.lua",
+		function(a_Body, a_Data)
+			if (a_Body) then
+				cUpdater:ParsePluginInfo(a_Body)
 			else
-				LOGWARNING("Error while downloading newer worldedit version")
+				-- The request failed, schedule a retry
+				cRoot:Get():GetDefaultWorld():QueueTask(
+					function()
+						cUpdater:CheckForNewerVersion()
+					end
+				)
 			end
-			return
 		end
-		
-		-- Write the ZIP data to the file. The filename looks like this: "WorldEdit v<DisplayVersion>.zip"
-		local ZipFile = assert(io.open("Plugins/WorldEdit v" .. (g_LatestVersion or "_Unknown") .. ".zip", "wb"), "Failed to open \"Plugins/WorldEdit v" .. (g_LatestVersion or "_Unknown") .. ".zip\"")
-		ZipFile:write(Data)
-		ZipFile:close()
-		
-		LOGINFO(string.format("New WorldEdit version downloaded to %q", "Plugins/WorldEdit v" .. (g_LatestVersion or "_Unknown") .. ".zip"))
-	end,	
-}
-
-
-
-
-
-function cUpdater:CheckForNewerVersion()
-	cNetwork:Connect("raw.githubusercontent.com", 443, VersionCheckForNewerVersionerCallbacks)
+	)
 end
 
 
 
 
 
-function cUpdater:DownloadLatestVersion()
-	cNetwork:Connect("codeload.github.com", 443, NewVersionDownloaderCallbacks)
+function cUpdater:DownloadLatestVersion(a_DisplayVersion)
+	if (cUpdater.s_DownloadAttempts > g_Config.Updates.NumAttempts) then
+		LOGWARNING("Error while downloading newer worldedit version")
+		return
+	end
+	cUpdater.s_DownloadAttempts = cUpdater.s_DownloadAttempts + 1;
+	
+	cUrlClient:Get("https://raw.githubusercontent.com/cuberite/WorldEdit/zip/master", 
+		function (a_Body, a_Data)
+			if (not a_Body or (a_Body:len() ~= tonumber(a_Data["Content-Length"]))) then
```

Comment:
It should work - the content-length is used (if the response isn't chunked) to distinguish the response's end and the potential beginning of another response over the same connection (we don't use that yet, though), so if the server cuts the answer short, the parser should report an error.